components/raftstore/src/store/peer.rs

```diff
@@ -3823,6 +3909,84 @@ where
         Ok(prs)
     }
 
+    pub fn maybe_witness_transfer_leader<T>(&mut self, ctx: &PollContext<EK, ER, T>) {
+        if !(self.is_witness() && self.is_leader()) {
+            return;
+        }
+        fail_point!("witness_transfer_leader", |_| {});
+        info!(
+            "witness try to transfer leader";
+            "region_id" => self.region_id,
+            "peer_id" => self.peer.get_id(),
+        );
+
+        if self.raft_group.raft.lead_transferee.is_some() {
+            info!("witness already transferring leader");
+            // already being in transfer leader
+            return;
+        }
+
+        let prs = self.raft_group.raft.prs();
+        // find other peers that are not down
+        // TODO: fitler out non-witnesses waiting for data
+        let (_, peer) = self
+            .region()
+            .get_peers()
+            .iter()
+            .filter(|peer| peer.id != self.peer.id)
+            .filter(|peer| !peer.is_witness && peer.role != PeerRole::Learner)
+            .filter(|peer| {
+                if let Some(instant) = self.peer_heartbeats.get(&peer.get_id()) {
+                    let elapsed = instant.saturating_elapsed();
+                    if elapsed < ctx.cfg.max_peer_down_duration.0 {
+                        return true;
+                    }
+                } else {
+                    info!("can't find peer heartbeat"; "peer_id" => peer.get_id());
+                }
+                false
+            })
+            .fold((0, None), |(max_matched, chosen), peer| {
```

Comment:
I have question:

Is it possible that after the selected peer is elected, when the witness leader replicates the log to this peer, the peer reject and prompts the witness leader to send a snapshot (but witness leader cann't gen snapshot)?  I'm not sure can the current filter condition + max_matched filter out such peers?

Reply:
why would the peer prompt send a snapshot?

Reply:
I mean is it possible that the elected peer is still more behind than the witness leader? So when it receives `MsgAppend` from the witness leader, it is rejected.

Reply:
Or am I misunderstanding the implementation of raft transfer leader?

Reply:
`the elected peer is still more behind than the witness leader`, it won't happen as raft log gc logic promises the witness would have the log.